app/models/stage.rb

```diff
@@ -254,4 +259,12 @@ def append_new_command
     previous = stage_commands.map(&:position).max || 0
     stage_commands.build(command: new_command, position: previous + 1)
   end
+
+  def environment_lock?(lock)
+    lock.resource_type == "Environment" && environments.any? { |e| lock.resource_equal(e) }
+  end
+
+  def project_lock?(lock)
+    lock.resource_type == "Project" && lock.resource_equal(project)
+  end
```

Comment:
this feels kinda wrong ... locking logic is now all over the place :(

Reply:
Yeah... I was on the fence about this. It kind of smells bad, but as you say it allows for locks to be used in plugins and it moves model-specific logic to the models.